File: apps/api/src/models/session.py
"""Session domain models."""

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class SessionStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class MessageRole(StrEnum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


class Message(BaseModel):
    role: MessageRole
    content: str
    tool_calls: list[dict[str, Any]] | None = None
    tool_call_id: str | None = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class SessionBase(BaseModel):
    agent_id: UUID
    task: str = Field(min_length=1, max_length=10_000)
    context: dict[str, Any] = Field(default_factory=dict)


class SessionCreate(SessionBase):
    pass


class Session(SessionBase):
    id: UUID
    user_id: UUID
    status: SessionStatus = SessionStatus.PENDING
    messages: list[Message] = Field(default_factory=list)
    result: dict[str, Any] | None = None
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: packages/types/src/session.ts
export type SessionStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export type MessageRole = "system" | "user" | "assistant" | "tool";

export interface Message {
  role: MessageRole;
  content: string;
  toolCalls?: Record<string, unknown>[];
  toolCallId?: string;
  timestamp: string;
}

export interface Session {
  id: string;
  userId: string;
  agentId: string;
  task: string;
  context: Record<string, unknown>;
  status: SessionStatus;
  messages: Message[];
  result: Record<string, unknown> | null;
  error: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface SessionCreate {
  agentId: string;
  task: string;
  context?: Record<string, unknown>;
}
